ion-schema/src/main/kotlin/com/amazon/ionschema/model/Constraint.kt

```diff
@@ -0,0 +1,236 @@
+package com.amazon.ionschema.model
+
+import com.amazon.ion.IonValue
+import com.amazon.ionschema.IonSchemaVersion
+import com.amazon.ionschema.internal.util.validateRegexPattern
+import kotlin.text.Regex as KRegex
+
+/**
+ * This internal only interface allows us to use an exhaustive `when` within this library.
+ * It will become `sealed` when we update to kotlin language version = 1.5 or higher.
```

Comment:
Mostly just because we haven't determined whether an upgrade would require a new major version, and we haven't had a particularly strong impetus before now.